File: 146-lru-cache/lru-cache.py
class Node:
    def __init__(self,key,val):
        self.key=key
        self.val=val
        self.prev=None
        self.next=None

class LRUCache:

    def __init__(self, capacity: int):
        self.cap=capacity
        self.hashmap={}

        self.left=ListNode(0,0)
        self.right=ListNode(0,0)
        self.left.next=self.right
        self.right.prev=self.left
    def remove(self,node):
        prev, nxt = node.prev, node.next
        prev.next, nxt.prev = nxt, prev

    def insert(self,node):
        prev, nxt = self.right.prev, self.right
        prev.next = nxt.prev = node
        node.next, node.prev = nxt, prev


    def get(self, key: int) -> int:
        if key in self.hashmap:
            self.remove(self.hashmap[key])
            self.insert(self.hashmap[key])
            return self.hashmap[key].val
        return -1        

        
        

    def put(self, key: int, value: int) -> None:
        if key in self.hashmap:
            self.remove(self.hashmap[key])
        self.hashmap[key]=Node(key,value)
        self.insert(self.hashmap[key])

        if len(self.hashmap) > self.cap:
            lru = self.left.next
            self.remove(lru)
            del self.hashmap[lru.key]


        

File: 146-lru-cache/lru-cache.java
class Node {
    int val;
    int key;
    Node prev;
    Node next;

    Node(int key , int val)
    {
        this.key = key;
        this.val = val;
    }
}

class LRUCache {
    private int cap;
    private Map<Integer , Node> cache ;
    private Node left;
    private Node right;


    public LRUCache(int capacity) {
        this.cap = capacity;

        this.cache = new HashMap<>();
        this.left = new Node(0 , 0);
        this.right = new Node(0 , 0);

        this.left.next = this.right ;
        this.right.prev = this.left ;


    }
    private void remove(Node node)
    {
        Node prev = node.prev;
        Node nxt = node.next;

        prev.next = nxt;
        nxt.prev = prev;
    }

    private void insert(Node node)
    {
        Node prev = this.right.prev;
        Node nxt = this.right;

        prev.next = node;
        nxt.prev = node;
        node.next = nxt;
        node.prev = prev;
    }
    public int get(int key) {

        if (cache.containsKey(key))
        {
            Node node = cache.get(key);
            remove(node);
            insert(node);
            return node.val;
        }
        return -1;

        
    }
    
    public void put(int key, int value) {
        if (cache.containsKey(key))
        {
            remove(cache.get(key));
        }
        Node node = new Node(key , value);
        cache.put(key , node);
        insert(node);
        if (cache.size() > cap)
        {
            Node lru = this.left.next;
            remove(lru);
            cache.remove(lru.key);
        }
        
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */